File: feign-core/src/main/java/name/abhijitsarkar/feign/core/model/FeignProperties.java
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3 of the License.
 *  *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of the GNU General Public License accompanies this software,
 * and is also available at http://www.gnu.org/licenses.
 *
 */

package name.abhijitsarkar.feign.core.model;

import lombok.Getter;
import name.abhijitsarkar.feign.persistence.DefaultIdGenerator;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.stereotype.Component;

import javax.annotation.PostConstruct;
import java.util.List;

import static java.lang.Boolean.FALSE;
import static java.lang.Boolean.TRUE;
import static java.util.Collections.emptyList;

/**
 * @author Abhijit Sarkar
 */
@Component
@ConfigurationProperties(prefix = "feign")
@Getter
@SuppressWarnings({"PMD.BeanMembersShouldSerialize", "PMD.SingularField"})
public class FeignProperties extends AbstractIgnorableRequestProperties {
    private RecordingProperties recording;
    private List<FeignMapping> mappings;
    private Delay delay;

    @PostConstruct
    public void postConstruct() {
        setMappings(mappings);
        setRecording(recording);
        setDelay(delay);
        setIgnoreCase(null);
        setIgnoreUnknown(null);
        setIgnoreEmpty(null);
    }

    public void setMappings(List<FeignMapping> mappings) {
        this.mappings = (mappings == null) ? emptyList() : mappings;
    }

    public void setRecording(RecordingProperties recording) {
        this.recording = (recording == null) ? new RecordingProperties() : recording;

        if (this.recording.getIdGenerator() == null) {
            this.recording.setIdGenerator(DefaultIdGenerator.class);
        }
    }

    public void setDelay(Delay delay) {
        this.delay = delay == null ? new Delay() : delay;
    }

    public void setIgnoreCase(Boolean ignoreCase) {
        super.setIgnoreCase((ignoreCase == null) ? FALSE : ignoreCase);
    }

    public void setIgnoreUnknown(Boolean ignoreUnknown) {
        super.setIgnoreUnknown((ignoreUnknown == null) ? TRUE : ignoreUnknown);
    }

    public void setIgnoreEmpty(Boolean ignoreEmpty) {
        super.setIgnoreEmpty((ignoreEmpty == null) ? TRUE : ignoreEmpty);
    }
}

File: feign-core/src/main/scala/name/abhijitsarkar/feign/core/model/FeignProperties.scala
/*
 * Copyright (c) 2016, the original author or authors.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of the GNU General Public License accompanies this software,
 * and is also available at http://www.gnu.org/licenses.
 */

package name.abhijitsarkar.feign.core.model

import java.lang.{Boolean => JavaBoolean}
import java.util.{Collections, List => JavaList}
import javax.annotation.PostConstruct

import name.abhijitsarkar.feign.persistence.DefaultIdGenerator
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.stereotype.Component

import scala.collection.JavaConverters._

/**
  * @author Abhijit Sarkar
  */
@Component
@ConfigurationProperties(prefix = "feign")
class FeignProperties extends AbstractIgnorableRequestProperties {
  var recording: RecordingProperties = _
  var mappings: JavaList[FeignMapping] = _

  @PostConstruct
  def postConstruct() {
    setMappings(mappings)
    setRecording(recording)
    setIgnoreCase(null.asInstanceOf[JavaBoolean])
    setIgnoreUnknown(null.asInstanceOf[JavaBoolean])
    setIgnoreEmpty(null.asInstanceOf[JavaBoolean])
  }

  def mappingsToScala = mappings.asScala

  def getMappings = mappings

  def setMappings(mappings: JavaList[FeignMapping]) {
    this.mappings = if (mappings == null) Collections.emptyList()
    else mappings
  }

  def getRecording = recording

  def setRecording(recording: RecordingProperties) {
    this.recording = if (recording == null) new RecordingProperties
    else recording

    if (this.recording.getIdGenerator == null)
      this.recording.setIdGenerator(classOf[DefaultIdGenerator])
  }

  override def setIgnoreCase(ignoreCase: JavaBoolean) {
    super.setIgnoreCase(if (ignoreCase == null) JavaBoolean.FALSE else ignoreCase)
  }

  override def setIgnoreUnknown(ignoreUnknown: JavaBoolean) {
    super.setIgnoreUnknown(if (ignoreUnknown == null) JavaBoolean.TRUE else ignoreUnknown)
  }

  override def setIgnoreEmpty(ignoreEmpty: JavaBoolean) {
    super.setIgnoreEmpty(if (ignoreEmpty == null) JavaBoolean.TRUE else ignoreEmpty)
  }

  override def toString = s"FeignProperties($recording, $mappings)"
}
